include/simdjson/ppc64/bitmask.h

```diff
@@ -3,11 +3,108 @@
 #endif // SIMDJSON_CONDITIONAL_INCLUDE
 
 namespace simdjson {
 namespace ppc64 {
-namespace {
+namespace bitmask {
+
+simdjson_constinit uint64_t ALL  = 0xFFFFFFFFFFFFFFFF;
+simdjson_constinit uint64_t NONE = 0xFFFFFFFFFFFFFFFF;
+simdjson_constinit uint64_t EVEN = 0x5555555555555555;
+simdjson_constinit uint64_t ODD  = 0xAAAAAAAAAAAAAAAA;
+
+// We sometimes call trailing_zero on inputs that are zero,
+// but the algorithms do not end up using the returned value.
+// Sadly, sanitizers are not smart enough to figure it out.
+SIMDJSON_NO_SANITIZE_UNDEFINED
+// This function can be used safely even if not all bytes have been
+// initialized.
+// See issue https://github.com/simdjson/simdjson/issues/1965
+SIMDJSON_NO_SANITIZE_MEMORY
+simdjson_inline int trailing_zeroes(uint64_t input_num) {
+#if SIMDJSON_REGULAR_VISUAL_STUDIO
+  unsigned long ret;
+  // Search the mask data from least significant bit (LSB)
+  // to the most significant bit (MSB) for a set bit (1).
+  _BitScanForward64(&ret, input_num);
+  return (int)ret;
+#else  // SIMDJSON_REGULAR_VISUAL_STUDIO
+  return __builtin_ctzll(input_num);
+#endif // SIMDJSON_REGULAR_VISUAL_STUDIO
+}
+
+/* result might be undefined when input_num is zero */
+simdjson_inline uint64_t clear_lowest_bit(uint64_t input_num) {
+  return input_num & (input_num - 1);
+}
+
+/* result might be undefined when input_num is zero */
+simdjson_inline int leading_zeroes(uint64_t input_num) {
+#if SIMDJSON_REGULAR_VISUAL_STUDIO
+  unsigned long leading_zero = 0;
+  // Search the mask data from most significant bit (MSB)
+  // to least significant bit (LSB) for a set bit (1).
+  if (_BitScanReverse64(&leading_zero, input_num))
+    return (int)(63 - leading_zero);
+  else
+    return 64;
+#else
+  return __builtin_clzll(input_num);
+#endif // SIMDJSON_REGULAR_VISUAL_STUDIO
+}
+
+#if SIMDJSON_REGULAR_VISUAL_STUDIO
+simdjson_inline int count_ones(uint64_t input_num) {
+  // note: we do not support legacy 32-bit Windows in this kernel
+  return __popcnt64(input_num); // Visual Studio wants two underscores
+}
+#else
+simdjson_inline int count_ones(uint64_t input_num) {
+  return __builtin_popcountll(input_num);
+}
+#endif
+
+simdjson_inline uint64_t add_carry_out(uint64_t value1, uint64_t value2, bool& carry_out) {
+#if SIMDJSON_REGULAR_VISUAL_STUDIO
+  uint64_t result = value1 + value2;
+  carry_out = result < value1;
+  return result;
+#else
+  unsigned long long result;
+  carry_out = __builtin_uaddll_overflow(value1, value2, &result);
+  return result;
+#endif
+}
+
+#if SIMDJSON_REGULAR_VISUAL_STUDIO
+using borrow_t = uint64_t;
+#else
+using borrow_t = unsigned long long;
+#endif
+
+simdjson_inline uint64_t subtract_borrow(const uint64_t value1, const uint64_t value2, borrow_t& borrow) noexcept {
+#if SIMDJSON_REGULAR_VISUAL_STUDIO
+  uint64_t result = value1 - value2 - borrow;
+  borrow_out = result >= value1;
+  return result;
+#else
+  return __builtin_subcll(value1, value2, borrow, &borrow);
```

Comment:
At a glance, it looks like __builtin_subcll is LLVM specific?

It might be worth guarding its usage:
https://gcc.gnu.org/onlinedocs/cpp/_005f_005fhas_005fbuiltin.html

It might worth examining alternatives:

https://godbolt.org/z/1WT9nPv6M

```C++


#include <cstdint>

using borrow_t = unsigned long long;

uint64_t subtract_borrow(const uint64_t value1, const uint64_t value2, borrow_t& borrow) noexcept {
   return __builtin_subcll(value1, value2, borrow, &borrow);
}

uint64_t subtract_borrow_manual(const uint64_t value1, const uint64_t value2, borrow_t& borrow) noexcept {
  uint64_t result = value1 - value2 - borrow;
  borrow = result >= value1;
  return result;
}
#if defined(_M_X64) || defined(__amd64__)
#include <x86intrin.h>

// visual studio has _subborrow_u64 in <intrin.h>
// https://learn.microsoft.com/en-us/cpp/intrinsics/x64-amd64-intrinsics-list?view=msvc-170
//
uint64_t subtract_borrow_intel(const uint64_t value1, const uint64_t value2, uint8_t& borrow) {
    uint64_t result;
    borrow = _subborrow_u64(borrow, value2, value1, (unsigned long long *)&result);
    return result;
}
#endif
```

Reply:
It's entirely possible; I did construct these to be analogues of the `add_overflow()` implementation for the given architecture (i.e. pulling from the same libraries and using the same #ifdefs)